docs/src/main/asciidoc/extension-authors-guide.adoc

```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
+the final target that is being produced.  For example, when running in developer mode the final output will not ask
+for GraalVM-specific build items such as `ReflectiveClassBuildItem`, so methods that only produce these
 items will not be invoked.
 
-Note that private methods and fields are not allowed, as injection is resolved at compile time via an annotation processor,
-and the resulting code does not have permission to inject private fields or invoke private methods.
+[id='consuming-values']
+==== Consuming values
 
-`BuildItem` instances should be immutable, as the producer/consumer model does not allow for mutation to be correctly
-ordered. This is not enforced but failure to adhere to this can result in race conditions.
+A build step may consume values from previous steps in the following ways:
+
+- By injecting a <<simple-build-items,simple build item>>
+- By injecting an `Optional` of a simple build item class
+- By injecting a `List` of a <<multi-build-items,multi build item>> class
+- By annotating the method with `@io.quarkus.deployment.annotations.Consumes`, giving the class name of a
+<<virtual-build-items,virtual build item>>
+
+Normally it is an error for a step which is included to consume a simple build item that is not produced by any other
+step.  In this way, it is guaranteed that all of the declared values will be present and non-`null` when a step is run.
+
+Sometimes a value isn't necessary for the build to complete, but might inform some behavior of the build step if it is
+present.  In this case, the value can be optionally injected.
+
+NOTE: Multi build values are always considered _optional_.  If not present, an empty list will be injected.
+
+[id='producing-weak-values']
+===== Weak value production
+
+Normally a build step is included whenever it produces a build item which is consumed by any other build step.  In this way,
+only the steps necessary to produce the final artifact(s) are included, and steps which pertain to extensions which are
+not installed or which only produce build items which are not relevant for the given artifact type are excluded.  Sometimes
+this is undesirable.
+
+For such cases, the `@io.quarkus.deployment.annotations.Weak` annotation may be used.  This annotation indicates that
+the build step should not automatically be included solely the basis of producing the annotated value.
+
+Certain types of build items are generally always consumed, such as generated classes or resources.
+An extension might produce a build item along with a generated class to facilitate the usage
+of that build item.  Such a build step would use the `@Weak` annotation on the generated class build item, while normally
+producing the other build item.  If the other build item is ultimately consumed by something, then the step would run
+and the class would be generated.  If nothing consumes the other build item, the step would not be included in the build
+process.
```

Comment:
I am totally lost on what is happening. Again I think something with a concrete example would help a lot to understand. Also what is this generated class? a bytecode generated one? And that one must be annotated with `@Weak`?